Add algorithmic description of deterministic signing algorithm

jonasnick opened the pull request:
Based on #8. Fixes #3. ~~WIP because it really should have test vectors.~~

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+'''''DeterministicSign(sk, othernonce, pk<sub>1..u</sub>, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m, rand)''''':
+* If ''len(rand) > 0:''
+** Let ''sk' '' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand)''
+* Else: let ''sk' = sk''
```
```suggestion
* Else:
** Let ''sk' = sk''
```
That's less ambiguous.

Oh I see it's just copied from above. Then I think we could do this here or we can do it in a separate PR.

jonasnick replied:
Added commit to change all offenders.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+* The public keys ''pk<sub>1..u</sub>'': ''u'' 32-byte arrays
+* The number ''v'' of tweaks with ''0 &le; v < 2^32''
+* The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
+* The tweak methods ''is_xonly_t<sub>1..v</sub>'' : ''v'' booleans
```
```suggestion
* The tweak methods ''is_xonly_t<sub>1..v</sub>'': ''v'' booleans
```

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be ommitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''othernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
```
```suggestion
''DeterministicSign'' requires the argument ''aggothernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
```

It's slightly longer but much clearer, I think. Otherwise "othernonce" sounds to me like a third kind of nonce (beside pub and sec)


----

> `output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.`

This sounds like the caller is required to run NonceAgg but in fact it's still possible to trust an aggregator with it.

jonasnick replied:
Ok, fixed this.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be ommitted if randomness is entirely unavailable.
```
```suggestion
Note that there are no optional arguments except ''rand'', which can be omitted if randomness is entirely unavailable.
```

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -147,25 +147,6 @@ Ordinary tweaking can be used to derive child public keys from an aggregate publ
-To obtain such a nonce generation algorithm ''NonceGen' '', the algorithm ''NonceGen'' should be modified as follows: The arguments ''sk'', ''aggpk'' and ''m'' are not optional and must be set precisely to the signer's secret key, the aggregate public key, and message of the session, respectively.
-In addition, ''NonceGen '' requires the ''pubnonce'' values of '''all''' other signers (concatenated in the order of signers), which can be provided via the ''extra_in'' argument.
-Hence, using ''NonceGen' '' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
-Further inputs can be added to ''extra_in'' as described in the [[#nonce-generation|Nonce Generation]] section.
-Lastly, if no randomness, not even low-quality randomness, is available, ''NonceGen' '' can be made deterministic by removing ''rand' '' and setting ''rand'' to ''sk''.
-Failure to provide the correct arguments to ''NonceGen' '' will allow attackers to extract secret keys.
```
Is this dropped intentionally?

jonasnick replied:
Yes, we mentioned this because we didn't have an algorithmic description. Now should be clear.

real-or-random replied:
Oh okay, yes. I got this wrong when reading it. I thought it's a separated change but it belonged to the "single deterministic signer" paragraph. So yeah, good that this is replaced by pseudocode. :)

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be ommitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''othernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
```
Right, it's very similar to the other "stateless" signer. 

I believe the following is true: If you want to make one signer stateless, it can't hurt to make that one deterministic.

So I suggest we unify this and only offer a single way to make a signer stateless, which is the deterministic one. I think an easy way to implement this in the document is to refer directly to the deterministic nonce gen in the "Nonce Generation" section.

jonasnick replied:
It hurts in the sense that it requires an implementation of a new algorithm `DeterministicSign` but the same can be achieved in a simpler way, by reusing the existing algorithms.

real-or-random replied:
Yeah, that's true. On the other hand, using `DeterministicSign` should be much safer the single signer, so I'm not sure if still worth to keep the added complexity of two variants. I think we should at least say that using `DeterministicSign` is highly recommend when a signer is made stateless.

jonasnick replied:
OTOH in the nonce modification section we warn against taking a noncegen change lightly. I added a sentence to refer to this section.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+''DeterministicSign'' requires the argument ''othernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
+
+==== Deterministic Signing ====
```
```suggestion
==== Deterministic and Stateless Signing for a Single Signer ====
```

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
+** Let ''sk' '' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand)''
+* Else: let ''sk' = sk''
+* Let aggpk = ''KeyAgg(pk<sub>1..u</sub>, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>)''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/stateless/nonce</sub>(sk' || othernonce || aggpk || m || bytes(1, i))) mod n'' for ''i = 1,2''
```
```suggestion
* Let ''k<sub>i</sub> = int(hash<sub>MuSig/deterministic/nonce</sub>(sk' || othernonce || aggpk || m || bytes(1, i))) mod n'' for ''i = 1,2''
```

I think if we call the algorithm DeterministicSign (which is a good name), that's slightly better. (It won't matter much though if we unify the two methods of stateless signing.)

real-or-random reviewed (commented):
Concept ACK

spelling out the algorithm is much easier to understand

jonasnick commented:
Probably easiest to get #9 in before we add test vectors here.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -412,6 +395,50 @@ The reference implementation is for demonstration purposes only and not to be us
+''DeterministicSign'' requires the argument ''aggothernonce'' which should be the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers (but can be provided by an untrusted party).
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
+
+==== Deterministic and Stateless Signing for a Single Signer ====
```
Maybe  add again a big warning here that this only works the signer who send the ''pubnonce'' last, for the people that skip to this heading without reading the text above it.

jonasnick replied:
Users also need to read above paragraph to understand what aggothernonce is, so I added a link from the aggothernonce argument to this section.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -412,6 +395,50 @@ The reference implementation is for demonstration purposes only and not to be us
+* Let ''R<sup>*</sup><sub>1</sub> = k<sub>1</sub>⋅G, R<sup>*</sup><sub>2</sub> = k<sub>2</sub>⋅G''
+* Let ''pubnonce = cbytes(R<sup>*</sup><sub>1</sub>) || cbytes(R<sup>*</sup><sub>2</sub>)''
+* Let ''secnonce = bytes(32, k<sub>1</sub>) || bytes(32, k<sub>2</sub>)''
+* Let ''aggnonce = NonceAgg(pubnonce, aggothernonce)''
```
This relies on the assumption that `NonceAgg(A, NonceAgg(B, C)) = NonceAgg(A, B, C)` but that's not true if `B = -C`.

Uff, I don't know. We could just require all individual other nonces here but that's not very elegant and it prevents the use of an aggregator.

Maybe the entire "optimization" is just making things worse and we should be able serialize infinity. Associativity and commutative are really cool properties. For example, what if you want to aggregate nonces in a tree-like structure? That's not possible either with the current NonceAgg. (Last time we removed algebraic structure was x-only keys and it didn't really make things easier....)

jonasnick replied:
In what way exactly does it rely on the assumption of associativity? If `B=-C` there's at least one disruptive signer who can prevent the scheme from succeeding anyway. As far as I can see, even though associativity doesn't hold, everything we care about still holds (i.e. we can identifying disruptive signers, ...). But of course that's more difficult to reason about...

real-or-random replied:
I believe "identifiable aborts" doesn't hold anymore.

The scenario I have in mind is (let's restrict our attention to the first group element in the nonce, suffices to explain the issue, all second group elements are computed honestly)
 - malicious signer A, sends nonce `A = aG` 
 - malicious signer B, sends nonce `B = -aG`
 - honest deterministic signer C, sends nonce `C`.

C computes `NonceAgg(C, NonceAgg(A, B)) = NonceAgg(C, G)  = C+G` and some final nonce `R`. Now A and B are able to produce correct partial sigs because they know `a`. But the final signature won't verify (it matches nonce `C` but not `C+G`).

I think one way out for C may be to blame everyone else in this case (or always blame the first signer that is not C). But that adds complexity again, and I'm not sure if it worth the hassle.

Another issue in this scenario is that A and B could use an honest aggregator. In that case, the aggregator really doesn't know who to blame because everyone has sent correct partial sigs. Blaming all signers is clearly wrong because C is honest.

But I think the issue is actually deeper, and exists already without deterministic signer. Assume a similar scenario:
 - malicious signer A, sends nonce `A = aG` 
 - malicious signer B, sends nonce `B = -aG`
 - honest aggregator

Aggregator computes `NonceAgg(A, B) = G`.

Here the honest aggregator can't tell who is to blame because partial sigs verify. Again, I believe then a simple fix would be to blame every signer, or blame always the first signer. You could also argue that we don't care about this case because the only honest party is the aggregator but I think that's a stretch. 

Another pro for associativity is just more flexibility. You may want to serialize intermediate state of nonce aggregation, i.e. you receive a bunch of bunches, call NonceAgg, store the serialized result in memory, and later add your own nonce. Or multiple aggregators in a tree (because you do nested MuSig, the simple variant where all signers know all other signers but still you want to do nonce aggregation). I think I would intuitively except associativity as a user, and I more and more tend to believe that serializing infinity is a simpler and less surprising solution than working around the issue. It's as simple as as using 33 zero bytes.

jonasnick replied:
All good points. Thanks, I see now. Ok, let's allow nonceagg to output infinity.

jonasnick replied:
We've since merged a PR that does this (#21).

jonasnick commented:
Will wait to write test vectors until #21 and #24 are merged.

jonasnick commented:
I rebased this PR which required taking into account #24 and #15, in particular. Also added a `deterministic_sign` function to the reference code and test vectors. There are some in-flight PRs affect this PR, including #50, #46 and #37 (which requires regenerating the test vectors). Anyway, this is ready for review.

real-or-random commented:
needs rebase again... :/ )I'll review soon, independent of rebased status.)

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -468,6 +451,53 @@ The reference implementation is for demonstration purposes only and not to be us
+First, on systems where obtaining uniformly random values is much harder than maintaining a global atomic counter, it can be beneficial to modify ''NonceGen''.
+Instead of drawing ''rand' '' uniformly at random, ''rand' '' can be the value of an atomic counter that is incremented whenever it is read.
+With this modification, the secret signing key ''sk'' of the signer generating the nonce is '''not''' an optional argument and must be provided to ''NonceGen''.
+The security of the resulting scheme is then depending on the requirement that the counter must never return the same output in two ''NonceGen'' invocations with the same ''sk''.
```
nit, slightly cleaner:
```suggestion
The security of the resulting scheme is then depending on the requirement that reading the counter must never yield the same counter value in two ''NonceGen'' invocations with the same ''sk''.
```

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -468,6 +451,53 @@ The reference implementation is for demonstration purposes only and not to be us
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be omitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''aggothernonce'' which should be the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers (but can be provided by an untrusted party).
```
```suggestion
''DeterministicSign'' requires the argument ''aggothernonce'' which should be set to the output of ''NonceAgg'' run on the ''pubnonce'' value of '''all''' other signers (but can be provided by an untrusted party).
```

real-or-random commented on bip-musig2/reference.py:
```diff
@@ -708,9 +801,10 @@ def test_sign_and_verify_random(iters):
-        secnonce_2 = bytes(64)
+        if i % 2 == 0:
+            psig_2 = sign(secnonce_2, sk_2, session_ctx)
+            # Clear the secnonce after use
+            secnonce_2 = bytes(64)
```
Does this really work? I can't imagine that this is reliable in a language like Python, and I don't think we should give readers the impression that this works. (But let's resolve this in another PR maybe. I see it's consistent with some other comments that we already have somewhere else in the reference code.)

real-or-random commented on bip-musig2/reference.py:
```diff
@@ -345,6 +345,45 @@ def sign(secnonce: bytes, sk: bytes, session_ctx: SessionContext) -> bytes:
+    assert R_2_ is not None
+    pubnonce = cbytes(R_1_) + cbytes(R_2_)
+    secnonce = bytes_from_int(k_1) + bytes_from_int(k_2)
+    try:
+        aggnonce = nonce_agg([pubnonce, aggothernonce])
+    except Exception:
+        raise InvalidContributionError(None, "aggothernonce")
```
```suggestion
    aggnonce = nonce_agg([pubnonce, aggothernonce])
```

jonasnick replied:
This would mean that the exception is `InvalidContributionError(1, "pubnonce")` but `pubnonce` is not an argument to `deterministic_sign`.

real-or-random replied:
indeed. nevermind.

jonasnick commented:
rebased

real-or-random reviewed (approved):
ACK 74f766fbda417eefaa4b8ce2900a8eeb0533871e

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,53 @@ The reference implementation is for demonstration purposes only and not to be us
+* Let ''R<sup>*</sup><sub>1</sub> = k<sub>1</sub>⋅G, R<sup>*</sup><sub>2</sub> = k<sub>2</sub>⋅G''
+* Let ''pubnonce = cbytes(R<sup>*</sup><sub>1</sub>) || cbytes(R<sup>*</sup><sub>2</sub>)''
+* Let ''secnonce = bytes(32, k<sub>1</sub>) || bytes(32, k<sub>2</sub>)''
+* Let ''aggnonce = NonceAgg(pubnonce, aggothernonce)''
```
Nit: add extra parens (to indicate tuple), say "fail if that fails"?

```suggestion
* Let ''aggnonce = NonceAgg((pubnonce, aggothernonce)); fail if that fails''
```

jonasnick replied:
Good catch, done.

robot-dreams commented on bip-musig2/reference.py:
```diff
@@ -361,6 +361,45 @@ def sign(secnonce: bytes, sk: bytes, session_ctx: SessionContext) -> bytes:
+    buf += len(msg).to_bytes(8, 'big')
+    buf += msg
+    buf += i.to_bytes(1, 'big')
+    return int_from_bytes(tagged_hash('MuSig/deterministic/nonce', buf))
```
This shouldn't matter except with negligible probability, but since the spec has an additional `mod n`, should that be added here as well?

jonasnick replied:
Yes, definitely. The regular nonce_gen algorithm misses this as well. I'll open a PR that fixes this.

robot-dreams commented on bip-musig2/det_sign_vectors.json:
```diff
@@ -0,0 +1,117 @@
+        "F95466D086770E689964664219266FE5ED215C92AE20BAB5C9D79ADDDDF3C0CF",
+        "2626262626262626262626262626262626262626262626262626262626262626262626262626"
+    ],
+    "valid_test_cases": [
```
Should we get a `rand = None` test case as well? For example (this is just the first test case with `rand` replaced):

```
        {
            "rand": null,
            "aggothernonce": "0337C87821AFD50A8644D820A8F3E02E499C931865C2360FB43D0A0D20DAFE07EA0287BF891D2A6DEAEBADC909352AA9405D1428C15F4B75F04DAE642A95C2548480",
            "key_indices": [0, 1, 2],
            "tweaks": [],
            "is_xonly": [],
            "msg_index": 0,
            "signer_index": 0,
            "expected": [
                "031BD2902D80BEE01DFE044B33D09906D58384C44E87E25F49D00A91FF509503A8028F4CB0195702628594C7778D7AD55E9EC1EB650D9AE88DE6011CCC620EFEF6BF",
                "7228033A3D51824EA3224DE4F941011FB0AB2F4BA5F824B7228AB2A670A0F9E4"
            ]
        },
```

jonasnick replied:
Yes, added a test.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,53 @@ The reference implementation is for demonstration purposes only and not to be us
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be omitted if randomness is entirely unavailable.
```
Wording nit:

```suggestion
Note that the only optional argument is ''rand'', which can be omitted if randomness is entirely unavailable.
```

jonasnick replied:
done

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,53 @@ The reference implementation is for demonstration purposes only and not to be us
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be omitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''aggothernonce'' which should be set to the output of ''NonceAgg'' run on the ''pubnonce'' value of '''all''' other signers (but can be provided by an untrusted party).
```
Just double checking, what was the motivation for requiring this to be already aggregated in the input, instead of taking `pubnonce<sub>2..u</sub>` as inputs?

real-or-random replied:
Because NonceAgg can be performed by an untrusted party/coordinator.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -139,6 +139,7 @@ In general, MuSig2 signers are stateful in the sense that they first generate ''
 However, it is possible for one of the signers to be stateless.
 This signer waits until it receives the ''pubnonce'' of all the other signers and until session parameters such as a message to sign, public keys, and tweaks are determined.
 Then, the signer can run ''NonceGen'', ''NonceAgg'' and ''Sign'' in sequence and send out its ''pubnonce'' along with its partial signature.
+Stateless signers may want to consider signing deterministically (see [[#modifications-to-nonce-generation|Modifications to Nonce Generation]]) to remove the reliance on the random number generator in the ''NonceGen'' algorithm.
```
- Would it make sense to say "Stateless signers **that don't have access to a secure random number generator**"?
- Should we make it more clear up front that only one signer can be made deterministic?

jonasnick replied:
1. I think that signers who _do have_ access to a secure RNG may also want to sign deterministically because it protects them if the RNG turns out to be not so secure.
2. The paragraph says "However, it is possible for one of the signers to be stateless. [...] Stateless signers may want to consider signing deterministically (see [[#modifications-to-nonce-generation|Modifications to Nonce Generation]])" which seems to already indicate that this only applies to a single signer.

robot-dreams replied:
Oops, I missed the non-highlighted part

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,53 @@ The reference implementation is for demonstration purposes only and not to be us
+With this modification, the secret signing key ''sk'' of the signer generating the nonce is '''not''' an optional argument and must be provided to ''NonceGen''.
+The security of the resulting scheme is then depending on the requirement that reading the counter must never yield the same counter value in two ''NonceGen'' invocations with the same ''sk''.
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
```
Just double checking, are there any API / clarity benefits to removing the `rand` argument altogether, and making it not need any randomness at all?

jonasnick replied:
Not sure if I understand you correctly, but I wouldn't want to remove the `rand` argument because it provides sidechannel resistance.

robot-dreams replied:
My question was, "is the convenience of removing an extra decision for the user (should I pass in `rand` or not) worth the security benefit of sometimes having `rand`". Sounds like the answer is "no", which I'm good with.

jonasnick replied:
Correct, I don't think that's worth it since `rand` can provide additional protection even if you don't trust it to be 100% secure.

robot-dreams reviewed (commented):
Looks great overall!

- Since I missed some of the discussions around this, I want to make sure I understand correctly: is the key idea that `DeterministicSign` needs to prevent the possibility of nonce reuse when another signer changes their public nonce?

- Do you think it's worth summarizing the guidance for which algorithm to use? For example:

|   | Secure RNG | Insecure or No RNG |
| ------------- | ------------- | ------------- |
| **Can Keep State**  | Default `NonceGen`  | `NonceGen` with atomic counter  |
| **Must Be Stateless**  | Default `NonceGen`  | `DeterministicSign`  |

real-or-random commented:
I think a table is a very nice idea but it should incorporate that deterministic signing is possible only for last signer. And I would rather state the requirements in the body of the table.

Two mockups:

<table>
<tr>
	<td>NonceGen
	<td>NonceGen with counter
	<td>DeterministicNonceGen
<tr> 
	<td> if randomness is good and state can be kept securely
	<td> if counter does not repeat and state can be kept securely
	<td> if used by last signer
</table>

<table>
<tr>
	<td>
	<td> needs secure randomness
	<td> needs secure counter
	<td> needs to keep state securely
    <td> needs aggregated nonce of all other signers (=can only be used by last signer)
<tr>
	<td> NonceGen
	<td> X
	<td>
	<td> X
    <td>
<tr>
	<td>NonceGen with counter
	<td>
	<td>X
	<td>X
    <td>
<tr>
	<td>DeterministicNonceGen
	<td>
	<td>
	<td>
	<td> X
</table>

jonasnick commented:
I like the idea to add a table. Added one based on @real-or-random's second mockup.

@robot-dreams
> is the key idea that DeterministicSign needs to prevent the possibility of nonce reuse when another signer changes their public nonce?

Yes, very other "session parameter" that could change is already accounted for in `NonceGen` (assuming the optional args aggpk and message are provided).

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,64 @@ The reference implementation is for demonstration purposes only and not to be us
+|-
+| CounterNonceGen       || &nbsp; || ✓      || ✓      || &nbsp;
+|-
+| DeterministicNonceGen || &nbsp; || &nbsp; || &nbsp; || ✓
```
Nit: Should this say `DeterministicSign` since that's what it's called below?

jonasnick replied:
done

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -471,6 +454,64 @@ The reference implementation is for demonstration purposes only and not to be us
+|-
+!                       !! needs secure randomness !! needs secure counter !! needs to keep state securely !! needs aggregated nonce of all other signers (only possible for one signer)
+|-
+| NonceGen              || ✓      || &nbsp; || ✓      || &nbsp;
```
Nit: Should the cells along the left side be headers as well, e.g. in this style: https://www.mediawiki.org/wiki/Help:Tables#Accessibility_of_table_header_cells

jonasnick replied:
Good idea, gave it a try.

robot-dreams reviewed (commented):
ACK 1915c41f397c4ce9e4ce568927610a1b9f515ad5

real-or-random reviewed (approved):
ACK 7427ef9f5e7f653ea8cc76898abd8abe49d7187c

robot-dreams commented:
ACK 7427ef9

Though did you intend to remove the `det_hash % n` with your latest force-push? If not, pre-emptive ACK on a commit that puts the `% n` back as well :)

jonasnick commented:
Oops. Sorry. I added the `det_hash % n` again.
